subsys/net/lib/mqtt_sn/mqtt_sn.c

```diff
@@ -643,6 +643,128 @@ static void mqtt_sn_do_ping(struct mqtt_sn_client *client)
 	}
 }
 
+static void mqtt_sn_do_will_topic_update(struct mqtt_sn_client *client)
+{
+	struct mqtt_sn_param p = {.type = MQTT_SN_MSG_TYPE_WILLTOPICUPD};
+
+	if (client == NULL) {
+		return;
+	}
+
+	if (client->state != MQTT_SN_CLIENT_ACTIVE) {
+		LOG_ERR("Cannot update will topic: not connected");
+		return;
+	}
+
+	LOG_INF("Updating will topic");
+
+	p.params.willtopicupd.topic.data = client->will_topic.data;
+	p.params.willtopicupd.topic.size = client->will_topic.size;
+	p.params.willtopicupd.retain = client->will_retain;
+	p.params.willtopicupd.qos = client->will_qos;
+
+	encode_and_send(client, &p, 0);
+}
+
+static void mqtt_sn_do_will_message_update(struct mqtt_sn_client *client)
+{
+	struct mqtt_sn_param p = {.type = MQTT_SN_MSG_TYPE_WILLMSGUPD};
+
+	if (client == NULL) {
+		return;
+	}
+
+	if (client->state != MQTT_SN_CLIENT_ACTIVE) {
+		LOG_ERR("Cannot update will message: not connected");
+		return;
+	}
+
+	LOG_INF("Updating will message");
+
+	p.params.willmsgupd.msg.data = client->will_msg.data;
+	p.params.willmsgupd.msg.size = client->will_msg.size;
+
+	encode_and_send(client, &p, 0);
+}
+
+static int process_will_topic_update(struct mqtt_sn_client *client, int64_t *next_cycle)
+{
+	const int64_t now = k_uptime_get();
+	int64_t next_attempt;
+
+	if (!client->will_topic_update.in_progress) {
+		return 0;
+	}
+
+	if (now == 0) {
+		next_attempt = 1;
+	} else if (client->will_topic_update.last_attempt == 0) {
+		next_attempt = 0;
+	} else {
+		next_attempt = client->will_topic_update.last_attempt + T_RETRY_MSEC;
+	}
+
+	if (next_attempt <= now) {
+		if (!client->will_topic_update.retries--) {
```

Comment:
Please do
```
if (client->will_topic_update.retries-- == 0) {
```

Reply:
Are you sure you want to have the `--` post decrement? When this case is `true`, the decrement will still be done, and leave `retries` as `UINT8_MAX`.

Reply:
@M1cha will retries wrap cause problems?

Reply:
Or simply decrement before the `if` statement?